presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -797,7 +797,15 @@ private Object nullInputEvaluator(final Collection<Symbol> nullSymbols, Expressi
         public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Expression> context)
         {
             Expression inheritedPredicate = context.get();
+            if (!extractConjuncts(inheritedPredicate).contains(node.getSemiJoinOutput().toSymbolReference())) {
+                return visitNonFilteringSemiJoin(node, context);
```

Comment:
IMO `visitNonFilteringSemiJoin` should be called for both cases, then possibly in `visitFilteringSemiJoin` you might have easier code:

```
Plan rewritten = visitNonFilteringSemiJoin(node, context);
if (extractConjuncts(inheritedPredicate).contains(node.getSemiJoinOutput().toSymbolReference())) {
    rewritten = visitFilteringSemiJoin(node, context);
}
return rewritten;
```

Reply:
It could allow to get rid of  common part:
```java
            // Push inheritedPredicates down to the source if they don't involve the semi join output
            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(inheritedPredicate)) {
                Expression rewrittenConjunct = allInference.rewriteExpressionAllowNonDeterministic(conjunct, in(sourceSymbols));
                // Since each source row is reflected exactly once in the output, ok to push non-deterministic predicates down
                if (rewrittenConjunct != null) {
                    sourceConjuncts.add(rewrittenConjunct);
                }
                else {
                    postJoinConjuncts.add(conjunct);
                }
            }
```

Yet it does not come at the cost.
 Note that `visitNonFilteringSemiJoin` can return either `FilterNode` or `SemiJoinNode`. This then would need to be handled here (or inside `visitFilteringSemiJoin`). Given the fact that common snippet is not that long I am not sure if that is worth an effort. This way at least shape is simple (though I agree there is some duplication).